compio-driver/src/fd.rs

```diff
@@ -0,0 +1,171 @@
+#[cfg(unix)]
+use std::os::fd::FromRawFd;
+#[cfg(windows)]
+use std::os::windows::io::{
+    FromRawHandle, FromRawSocket, OwnedHandle, OwnedSocket, RawHandle, RawSocket,
+};
+use std::{
+    future::{poll_fn, Future},
+    mem::ManuallyDrop,
+    panic::RefUnwindSafe,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc,
+    },
+    task::Poll,
+};
+
+use futures_util::task::AtomicWaker;
+
+use crate::{AsRawFd, OwnedFd, RawFd};
+
+#[derive(Debug)]
+struct Inner {
+    fd: OwnedFd,
+    waits: AtomicUsize,
+    waker: AtomicWaker,
+}
+
+impl RefUnwindSafe for Inner {}
+
+/// A shared fd. It is passed to the operations to make sure the fd won't be
+/// closed before the operations complete.
+#[derive(Debug, Clone)]
+pub struct SharedFd(Arc<Inner>);
+
+impl SharedFd {
+    /// Create the shared fd from an owned fd.
+    pub fn new(fd: impl Into<OwnedFd>) -> Self {
+        Self(Arc::new(Inner {
+            fd: fd.into(),
+            waits: AtomicUsize::new(0),
+            waker: AtomicWaker::new(),
+        }))
+    }
+
+    /// Try to take the inner owned fd.
+    pub fn try_unwrap(self) -> Result<OwnedFd, Self> {
+        let this = ManuallyDrop::new(self);
+        if let Some(fd) = unsafe { Self::try_unwrap_inner(&this) } {
+            Ok(fd)
+        } else {
+            Err(ManuallyDrop::into_inner(this))
+        }
+    }
+
+    // SAFETY: if `Some` is returned, the method should not be called again.
+    unsafe fn try_unwrap_inner(this: &ManuallyDrop<Self>) -> Option<OwnedFd> {
+        let ptr = ManuallyDrop::new(std::ptr::read(&this.0));
+        // The ptr is duplicated without increasing the strong count, should forget.
+        match Arc::try_unwrap(ManuallyDrop::into_inner(ptr)) {
+            Ok(inner) => Some(inner.fd),
+            Err(ptr) => {
+                std::mem::forget(ptr);
+                None
+            }
+        }
+    }
+
+    /// Wait and take the inner owned fd.
+    pub fn take(self) -> impl Future<Output = Option<OwnedFd>> {
+        let this = ManuallyDrop::new(self);
+        async move {
+            poll_fn(|cx| {
```

Comment:
Remove the `async move` and add the `move` to the closure. `PollFn` is already a future and the closure can keep `this`.